### PR TITLE
support parse opengauss create index

### LINKED ISSUE #27862: Support parsing openGauss  CREATE sql
# Background
Hi community,

ShardingSphere parser engine helps users parse a SQL to get the AST (Abstract Syntax Tree) and visit this tree to get SQLStatement (Java Object). Currently, we are planning to enhance the support for openGauss SQL parsing in ShardingSphere.

More details:
https://shardingsphere.apache.org/document/current/en/reference/sharding/parse/

# Task
This issue is to support more openGauss sql parse, as follows:
```sql
 CREATE INDEX ds_customer_address_p1_index2 ON tpcds.customer_address_p1(CA_ADDRESS_SK) LOCAL
(
    PARTITION CA_ADDRESS_SK_index1,
    PARTITION CA_ADDRESS_SK_index2 TABLESPACE example3,
    PARTITION CA_ADDRESS_SK_index3 TABLESPACE example4
) 
TABLESPACE example2;

openGauss=CREATE INDEX ds_customer_address_p1_index3 ON tpcds.customer_address_p1(CA_ADDRESS_ID) GLOBAL;

openGauss=CREATE INDEX ds_customer_address_p1_index4 ON tpcds.customer_address_p1(CA_ADDRESS_ID);

```

```sql
 ALTER INDEX tpcds.ds_customer_address_p1_index2 MOVE PARTITION CA_ADDRESS_SK_index2 TABLESPACE example1;

```

```sql
 ALTER INDEX tpcds.ds_customer_address_p1_index2 MOVE PARTITION CA_ADDRESS_SK_index3 TABLESPACE example2;

```

```sql
 ALTER INDEX tpcds.ds_customer_address_p1_index2 RENAME PARTITION CA_ADDRESS_SK_index1 TO CA_ADDRESS_SK_index4;

```

```sql
 create table cgin_create_test(a int, b text) with (orientation = column);

```

# Process
1. First confirm that this is a correct openGauss sql syntax, if not please leave a message under the issue and ignore it;
2. Compare SQL definitions in Official SQL Doc and ShardingSphere SQL Doc;
3. If there is any difference in ShardingSphere SQL Doc, please correct them by referring to the Official SQL Doc;
4. Run mvn install the current_file_module;
5. Check whether there are any exceptions. If indeed, please fix them. (Especially xxxVisitor.class);
6. Add new corresponding SQL case in SQL Cases and expected parsed result in Expected Statement XML;
7. Run SQLParserParameterizedTest to make sure no exceptions.

# Relevant Skills
1. Master JAVA language
2. Have a basic understanding of Antlr `g4` file
3. Be familiar with openGauss SQLs

### PR DESCRIPTION
**Pull request by tmyVvit**
Fixes #27862

Changes proposed in this pull request:
- Support parse openGauss create index with partition and tablespace sql
- Support parse openGauss alter index move partition sql
- Support parse openGauss alter index rename partition sql
---

Before committing this PR, I'm sure that I have checked the following options:
- [x] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [x] I have self-reviewed the commit code.
- [x] I have (or in comment I request) added corresponding labels for the pull request.
- [x] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [x] I have added corresponding unit tests for my changes.

### CODE REVIEW

**Comment by tmyVvit** [comment]
hello @strongduanmu, can you give advice how to fix the failing checks. I find the log shows zookeeper session closed. But I don't know it is anything wrong with the code changes or just the environment issue.
image

**Comment by strongduanmu** [comment]
Can you run GeneralDQLE2EIT on your local machine?

**Comment by tmyVvit** [comment]
I can run GeneralDQLE2EIT, and there is no error on my local

**Comment by strongduanmu** [comment]
> I can run GeneralDQLE2EIT, and there is no error on my local

You can modify it-env.props according to fail e2e test case.

**Comment by taojintianxia** [comment]
this PR will migrate to milestone 5.4.2

**Comment by strongduanmu** [comment]
Hi @tmyVvit, can you merge master branch to keep code consistent with the master branch?

**Comment by tmyVvit** [comment]
Sure! I will merge it tonight


Zhengqiang Duan ***@***.***>于2023年10月8日 周日13:59写道：

File: parser/sql/dialect/opengauss/src/main/antlr4/imports/opengauss/DDLStatement.g4 [hunk 1]
```diff
@@ -348,6 +361,7 @@ partitionCmd
 
 alterIndexDefinitionClause
     : renameIndexSpecification | alterIndexDependsOnExtension | alterIndexSetTableSpace | alterTableCmds | indexPartitionCmd
```
Comment (strongduanmu) [inline parser/sql/dialect/opengauss/src/main/antlr4/imports/opengauss/DDLStatement.g4:363]: Can you write the rules one by one? According to the code specification, when there are more than 5 branches, a rule needs to be written one line.

File: parser/sql/dialect/opengauss/src/main/java/org/apache/shardingsphere/sql/parser/opengauss/visitor/statement/OpenGaussStatementVisitor.java [hunk 2]
```diff
@@ -1407,4 +1420,60 @@ public ASTNode visitAttrs(final AttrsContext ctx) {
     public ASTNode visitSignedIconst(final SignedIconstContext ctx) {
         return new NumberLiteralValue(ctx.getText());
     }
+    
+    @SuppressWarnings("unchecked")
+    @Override
+    public ASTNode visitCreateIndexPartitionClause(final OpenGaussStatementParser.CreateIndexPartitionClauseContext ctx) {
+        IndexPartitionTypeEnum indexPartitionType = null;
+        if (ctx.LOCAL() != null || ctx.GLOBAL() != null) {
```
Comment (strongduanmu) [inline parser/sql/dialect/opengauss/src/main/java/org/apache/shardingsphere/sql/parser/opengauss/visitor/statement/OpenGaussStatementVisitor.java:1428]: Please put null codition on the left side.

File: parser/sql/dialect/opengauss/src/main/java/org/apache/shardingsphere/sql/parser/opengauss/visitor/statement/OpenGaussStatementVisitor.java [hunk 3]
```diff
@@ -1407,4 +1420,60 @@ public ASTNode visitAttrs(final AttrsContext ctx) {
...
+        }
+        IndexPartitionsSegment result = new IndexPartitionsSegment(ctx.getStart().getStartIndex(), ctx.getStop().getStopIndex(), indexPartitionType);
+        if (null != ctx.indexPartitionElemList()) {
+            result.getIndexPartitionSegment().addAll(((CollectionValue<IndexPartitionSegment>) visit(ctx.indexPartitionElemList())).getValue());
+        }
+        return result;
+    }
+    
+    @Override
+    public ASTNode visitIndexPartitionElemList(final OpenGaussStatementParser.IndexPartitionElemListContext ctx) {
```
Comment (strongduanmu) [inline parser/sql/dialect/opengauss/src/main/java/org/apache/shardingsphere/sql/parser/opengauss/visitor/statement/OpenGaussStatementVisitor.java:1439]: Please import subclass directly.

File: parser/sql/dialect/opengauss/src/main/java/org/apache/shardingsphere/sql/parser/opengauss/visitor/statement/OpenGaussStatementVisitor.java [hunk 4]
```diff
@@ -1407,4 +1420,60 @@ public ASTNode visitAttrs(final AttrsContext ctx) {
...
+        CollectionValue<IndexPartitionSegment> result = new CollectionValue<>();
+        for (OpenGaussStatementParser.IndexPartitionElemContext each : ctx.indexPartitionElem()) {
+            result.getValue().add((IndexPartitionSegment) visit(each));
+        }
+        return result;
+    }
+    
+    @Override
+    public ASTNode visitIndexPartitionElem(final OpenGaussStatementParser.IndexPartitionElemContext ctx) {
```
Comment (strongduanmu) [inline parser/sql/dialect/opengauss/src/main/java/org/apache/shardingsphere/sql/parser/opengauss/visitor/statement/OpenGaussStatementVisitor.java:1448]: Please import subclass directly.

File: parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/ddl/partition/PartitionNameSegment.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+@RequiredArgsConstructor
+@Getter
+public class PartitionNameSegment implements SQLSegment {
+    
+    private final int startIndex;
+    
+    private final int stopIndex;
+    
+    private final IdentifierValue identifier;
+
```
Comment (strongduanmu) [inline parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/ddl/partition/PartitionNameSegment.java:37]: Please remove useless blank line.

File: parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/opengauss/segment/IndexPartitionSegment.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.SQLSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.ddl.partition.PartitionNameSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.ddl.tablespace.TablespaceSegment;
+
+@RequiredArgsConstructor
+@Getter
+@Setter
+public class IndexPartitionSegment implements SQLSegment {
```
Comment (strongduanmu) [inline parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/opengauss/segment/IndexPartitionSegment.java:31]: Please add final for this class.

File: parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/opengauss/segment/IndexPartitionSegment.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    
+    private final int startIndex;
+    
+    private final int stopIndex;
+    
+    private final PartitionNameSegment partitionName;
+    
+    private TablespaceSegment tablespace;
+    
+    public Optional<TablespaceSegment> getTablespace() {
```
Comment (strongduanmu) [inline parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/opengauss/segment/IndexPartitionSegment.java:41]: Please add javadoc for public method.

File: parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/opengauss/segment/IndexPartitionTypeEnum.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.shardingsphere.sql.parser.sql.dialect.statement.opengauss.segment;
+
+public enum IndexPartitionTypeEnum {
```
Comment (strongduanmu) [inline parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/opengauss/segment/IndexPartitionTypeEnum.java:20]: Please add javadoc for this class.

File: parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/opengauss/segment/IndexPartitionsSegment.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.util.Collection;
+import java.util.LinkedList;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.ddl.AlterDefinitionSegment;
+
+@RequiredArgsConstructor
+@Getter
+public class IndexPartitionsSegment implements AlterDefinitionSegment {
```
Comment (strongduanmu) [inline parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/opengauss/segment/IndexPartitionsSegment.java:28]: Please add final and javadoc for this class.

File: parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/opengauss/segment/MovePartitionSegment.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.SQLSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.ddl.partition.PartitionNameSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.ddl.tablespace.TablespaceSegment;
+
+@RequiredArgsConstructor
+@Getter
+@Setter
+public class MovePartitionSegment implements SQLSegment {
```
Comment (strongduanmu) [inline parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/opengauss/segment/MovePartitionSegment.java:30]: Please add final and javadoc for this class.

File: parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/opengauss/segment/RenamePartitionSegment.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.SQLSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.ddl.partition.PartitionNameSegment;
+
+@RequiredArgsConstructor
+@Getter
+@Setter
+public class RenamePartitionSegment implements SQLSegment {
```
Comment (strongduanmu) [inline parser/sql/statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/opengauss/segment/RenamePartitionSegment.java:29]: Please add final and javadoc for this class.

File: test/it/parser/src/main/java/org/apache/shardingsphere/test/it/sql/parser/internal/asserts/statement/ddl/impl/AlterIndexStatementAssert.java [hunk 12]
```diff
@@ -67,4 +82,55 @@ private static void assertIndex(final SQLCaseAssertContext assertContext, final
...
+    
+    private static void assertMovePartition(final SQLCaseAssertContext assertContext, final AlterIndexStatement actual, final AlterIndexStatementTestCase expected) {
+        if (actual instanceof OpenGaussAlterIndexStatement) {
+            OpenGaussAlterIndexStatement ogActual = (OpenGaussAlterIndexStatement) actual;
+            if (null == expected.getMovePartition()) {
+                assertFalse(ogActual.getMovePartition().isPresent(), assertContext.getText("Actual move partition should not exist."));
+            } else {
+                assertTrue(ogActual.getMovePartition().isPresent(), assertContext.getText("Actual move partition should exist."));
+                MovePartitionSegment actualMovePartition = ogActual.getMovePartition().get();
+                // assert partition
```
Comment (strongduanmu) [inline test/it/parser/src/main/java/org/apache/shardingsphere/test/it/sql/parser/internal/asserts/statement/ddl/impl/AlterIndexStatementAssert.java:94]: Please remove useless inline comment.

**Review by strongduanmu — CHANGES_REQUESTED** [review]
Hi @tmyVvit, thank you for your great contribution. There are a few minor issues you need to take a look at.

**Comment by tmyVvit** [comment]
Hi @strongduanmu, thanks your suggestion, I have fixed the issues

**Comment by strongduanmu** [comment]
Since this pr has long time no response, I will close it. You can submit new pr when you are free.

**Pull request closed without merging**